### PR TITLE
Introduce alpha configuration option to enable testing of structured configuration

### PR DESCRIPTION
**Pull request by JoelSpeed**
## Description

This PR introduces two new flags `--alpha-config` and `--convert-config-to-alpha` and a new local testing environment.

The `--alpha-config` flag takes the path to a YAML structured configuration file. When set, this will be loaded instead of the normal configuration for the options that have been moved to the alpha configuration.

The `--convert-config-to-alpha` flag will read the existing config and print out the converted options, this should help ease migration for those who are trying out the alpha config.

## Motivation and Context

This will allow developers to test the structured configuration and options that we are converting.
It will also allow some early adopters to try out the new config and hopefully find bugs in the code before general release.

## How Has This Been Tested?

Manually, using the new test environment.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.
- [x] Add tests for loading yaml configuration
- [x] Add appropriate comments throughout new code in this PR

### CODE REVIEW

**Comment by JoelSpeed** [comment]
This will likely need to be reviewed commit by commit. To get the yaml config reading as I wanted I have to remove pointers from the HeaderValue struct, this touched a lot of files! I may make this a separate PR 🤔

**Comment by JoelSpeed** [comment]
I created #912 as a pre-factor PR, once that's merged the diff for this PR will look a lot cleaner

**Comment by JoelSpeed** [comment]
Implementation for this is done now, docs changes are still required but I will follow up in a secondary PR with those once I've sorted out the docs generator and moving that into its own repository

File: main.go [hunk 1]
```diff
@@ -74,3 +77,91 @@ func main() {
 	}()
 	s.ListenAndServe()
 }
+
```
Comment (NickMeves) [inline main.go:80]: These methods (and the associated test suite -- which is good) feel out of place here.

We had talked about cleaning up all the miscellaneous files & methods at the top of the project.

I don't think these belong under `pkg` and exposed publicly. Is this the opportunity to start using an `internal` area and getting some non-public packages setup there?

Options & configuration validation logic jump to my mind as good candidates for an `internal` placement.

File: main_test.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+			configContent:      testCoreConfig,
+			alphaConfigContent: testAlphaConfig,
+			expectedOptions:    testExpectedOptions,
+		}),
+		Entry("with bad legacy configuration", loadConfigurationTableInput{
+			configContent:   testCoreConfig + "unknown_field=\"something\"",
+			expectedOptions: func() *options.Options { return nil },
+			expectedErr:     errors.New("failed to load config: error unmarshalling config: 1 error(s) decoding:\n\n* '' has invalid keys: unknown_field"),
+		}),
+		Entry("with bad alpha configuration", loadConfigurationTableInput{
```
Comment (NickMeves) [inline main_test.go:202]: This looks like just a test case for invalid YAML. Do you have one where the YAML is valid but the underlying schema doesn't match what we expect? (e.g. array where string is expected)

Or do you have more rigorous test cases handled already in the underlying alpha config packages -- making it redundant here?

File: pkg/apis/options/load.go [hunk 3]
```diff
@@ -132,3 +135,28 @@ func isUnexported(name string) bool {
...
+	if configFileName == "" {
+		return errors.New("no configuration file provided")
+	}
+
+	data, err := ioutil.ReadFile(configFileName)
+	if err != nil {
+		return fmt.Errorf("unable to load config file: %w", err)
+	}
+
+	// UnmarhsalExact will return an error if the config includes options that are
```
Comment (NickMeves) [inline pkg/apis/options/load.go:155]:
```suggestion
	// UnmarshalExact will return an error if the config includes options that are
```

File: pkg/apis/options/load_test.go [hunk 4]
```diff
@@ -295,10 +297,187 @@ var _ = Describe("Load", func() {
...
+				expectedOutput: &TestOptions{},
+				expectedErr:    errors.New("no configuration file provided"),
+			}),
+			Entry("with invalid YAML", loadYAMLTableInput{
+				configFile:     []byte("\tfoo: bar"),
+				input:          &TestOptions{},
+				expectedOutput: &TestOptions{},
+				expectedErr:    errors.New("error unmarshalling config: error converting YAML to JSON: yaml: found character that cannot start any token"),
+			}),
+			Entry("with extra fields in the YAML", loadYAMLTableInput{
```
Comment (NickMeves) [inline pkg/apis/options/load_test.go:397]: Actually, maybe my comment from before is more appropriate here? A test case where the field is expected but the type doesn't match out schema.

File: pkg/apis/options/load_test.go [hunk 5]
```diff
@@ -295,10 +297,187 @@ var _ = Describe("Load", func() {
...
+injectResponseHeaders:
+- name: X-Secret
+  values:
+  - value: c2VjcmV0
+`)
+
+		By("Creating a config file")
+		configFile, err := ioutil.TempFile("", "oauth2-proxy-test-alpha-config-file")
+		Expect(err).ToNot(HaveOccurred())
+		defer configFile.Close()
```
Comment (NickMeves) [inline pkg/apis/options/load_test.go:434]: Gosec or IDEs might flag these for ignoring the error return. This is the pattern I tried to get everywhere to get us a clean Gosec scan report: https://github.com/oauth2-proxy/oauth2-proxy/blob/a21e80eb00ea067ab29970e6a053259d9bf93c64/oauthproxy_test.go#L337

Comment (NickMeves) [reply pkg/apis/options/load_test.go:434]: But truthfully, in looking around -- I see the standard `defer thing.Close()` pattern throughout other tests.

Gosec might've only been on non-test code? I can't recall. Or its fickle -- I guess we get what we pay for 😜

**Review by NickMeves — COMMENTED** [review]
As always, very clean, well tested code 👍 

I left a few questions on organization & a couple corner cases that might be good to get tests cases around in the test tables. Otherwise looks great.

File: main.go (hunk 1 above)
Comment (JoelSpeed) [reply main.go:80]: I'd happily move them to another file if you'd prefer, but I think making a package for this might end up being overkill as this should all go away in our long term plan right? Once we are structured config only this can all be deleted.

I'm not sure about the use of `internal`, how much do we care if other people are importing these bits of code? We aren't claiming this is a library, but if people want to use bits of the code I don't really mind (eg middleware for example). I could actually imagine someone importing the `Options` package to generate configuration or something so we probably don't want to make that internal either.

I feel like we could possibly have a big conversation about this 😅  I might read up on when people are using `internal`

File: main_test.go (hunk 2 above)
Comment (JoelSpeed) [reply main_test.go:202]: The intention here was just to have basic tests to show we can load yaml and if there's an error, then an error is thrown. Having all of the corner cases tested should be the responsibility of the unit tests for the `LoadYAML` method that this calls IMO, if I haven't got something there (which I don't think I have) I will add it

File: pkg/apis/options/load.go (hunk 3 above)
Comment (JoelSpeed) [reply pkg/apis/options/load.go:155]: Should actually be `UnmarshalStrict` now, changed yaml library and forgot to update the comment 😅

File: pkg/apis/options/load_test.go (hunk 4 above)
Comment (JoelSpeed) [reply pkg/apis/options/load_test.go:397]: Yep! Will add a case for this, I think for this you mean something like putting a list where we are expecting a string?

File: pkg/apis/options/load_test.go (hunk 5 above)
Comment (JoelSpeed) [reply pkg/apis/options/load_test.go:434]: I think gosec is disabled on test code, there's definitely a separate configuration for the test code linting at least 🤔  I don't mind having gosec complaints for test code if you're ok with it

File: pkg/apis/options/load_test.go (hunk 4 above)
Comment (NickMeves) [reply pkg/apis/options/load_test.go:397]: Perfect - exactly what I was thinkings. New tests cover it 👍

File: pkg/apis/options/load_test.go (hunk 5 above)
Comment (NickMeves) [reply pkg/apis/options/load_test.go:446]: Yeah, that's probably for the best to keep things simple for other contributors. I already wish they wrote more/better tests for PRs, so I don't want to raise the barrier to entry even higher

**Review by NickMeves — APPROVED** [review]
LGTM! Thanks for the added test cases.

**Comment by JoelSpeed** [comment]
This was rebased to fix the changelog conflict, otherwise no changes since the approval, merging
